benchmarks/gups/gups.cpp

```diff
@@ -0,0 +1,201 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+/*! \brief file gups.cpp
+
+    An implementation of something like HPCC RandomAccess.
+*/
+
+#include "Kokkos_Core.hpp"
+#include <cstdio>
+#include <cstdlib>
+#include <cmath>
+#include <chrono>
+#include <numeric>
+#include <algorithm>
+#include <random>
+
+#define HLINE "-------------------------------------------------------------\n"
+
+using Index = int;
+using Datum = int64_t;
+
+using DefaultHostMemorySpace = Kokkos::DefaultHostExecutionSpace::memory_space;
+using IndexView              = Kokkos::View<Index*>;
+using DataView               = Kokkos::View<Datum*>;
+
+using Clock    = std::chrono::steady_clock;
+using Duration = std::chrono::duration<double>;
+
+using RandomDevice = std::random_device;
+using RNG          = std::mt19937;
+
+IndexView randomized_indices(const Index indicesCount, const Index dataCount,
+                             RNG& rng) {
+  // generate random indices 0..dataCount
+  std::uniform_int_distribution<Index> uid(0, dataCount);
+  std::vector<Index> indices(indicesCount);
+  std::generate(indices.begin(), indices.end(), [&]() { return uid(rng); });
+
+  // Copy to the default space and return
+  Kokkos::View<Index*, DefaultHostMemorySpace, Kokkos::MemoryUnmanaged>
+      unmanaged_indices(indices.data(), indices.size());
+  IndexView dev_indices("dev_indices", indicesCount);
+  Kokkos::deep_copy(dev_indices, unmanaged_indices);
+  return dev_indices;
+}
+
+IndexView permuted_indices(const Index indicesCount, const Index dataCount,
+                           RNG& rng) {
+  // create a permutation array of offsets into the data
+  std::vector<Index> perm(dataCount);
+  std::iota(perm.begin(), perm.end(), 0);
+  std::shuffle(perm.begin(), perm.end(), rng);
+
+  // indices is repeated copies of the permutation array
+  // (or the first entries of the permutation array if there
+  // are fewer indices than data elements)
+  Kokkos::View<Index*, DefaultHostMemorySpace> indices("indices", indicesCount);
+  for (Index i = 0; i < Index(indices.extent(0)); ++i) {
+    indices(i) = perm[i % perm.size()];
+  }
+
+  // Copy to the default space and return
+  IndexView dev_indices("dev_indices", indicesCount);
+  Kokkos::deep_copy(dev_indices, indices);
+  return dev_indices;
```

Comment:
You could create a mirror view and copy, that would avoid duplicating the view when there is no "device" backend enabled.  You don't need to change.